python/ffsim/states/slater.py

```diff
@@ -0,0 +1,296 @@
+
+        if n == 0:
+            sampled_configuration = np.zeros((shots, norb), dtype=int)
+            sampled_configuration = np.ones((shots, norb), dtype=int)
+        elif n == norb:
+            sampled_configuration = np.ones((shots, norb), dtype=int)
+
+        else:
+            sampled_configuration = _sample_spinless_direct(rdm, shots, rng)
+
+        return convert_bitstring_type(
+            sampled_configuration[:, orbs],
+            BitstringType.BIT_ARRAY,
+            bitstring_type,
+            length=len(orbs),
+        )
+
+    else:
+        # Spinful case
+        rdm_a, rdm_b = rdm
+
+        if isinstance(nelec, tuple):
+            n_a, n_b = nelec
+        else:
+            n_a = nelec // 2
+            n_b = nelec // 2
+
+        n_a = round(np.real(np.sum(np.diag(rdm_a))))
+        n_b = round(np.real(np.sum(np.diag(rdm_b))))
+        norb, _ = rdm_a.shape
+
+        if orbs is None:
+            orbs_a = range(norb)  # np.arange(norb, dtype=int)
+            orbs_b = range(norb)  # np.arange(norb, dtype=int)
+        elif isinstance(orbs, list):
+            orbs_a = orbs
+            orbs_b = orbs
+        else:
+            orbs_a = orbs[0]
+            orbs_b = orbs[1]
+
+        if n_a == 0:
+            sampled_configuration_a = np.zeros((shots, norb), dtype=int)
+        elif n_a == norb:
+            sampled_configuration_a = np.ones((shots, norb), dtype=int)
+        else:
+            sampled_configuration_a = _sample_spinless_direct(rdm_a, shots, rng)
+
+        if n_b == 0:
+            sampled_configuration_b = np.zeros((shots, norb), dtype=int)
+        elif n_b == norb:
+            sampled_configuration_b = np.ones((shots, norb), dtype=int)
+        else:
+            sampled_configuration_b = _sample_spinless_direct(rdm_b, shots, rng)
+
+        if concatenate:
+            sampled_configuration = np.concatenate(
+                (
+                    sampled_configuration_b[:, orbs_b],
+                    sampled_configuration_a[:, orbs_a],
+                ),
+                axis=1,
+            )
+            return convert_bitstring_type(
+                sampled_configuration,
+                BitstringType.BIT_ARRAY,
+                bitstring_type,
+                length=len(orbs_a) + len(orbs_b),
+            )
+
+        return convert_bitstring_type(
+            sampled_configuration_b[:, orbs_b],
+            BitstringType.BIT_ARRAY,
+            bitstring_type,
+            length=len(orbs_b),
+        ), convert_bitstring_type(
+            sampled_configuration_a[:, orbs_a],
+            BitstringType.BIT_ARRAY,
+            bitstring_type,
+            length=len(orbs_b),
+        )
+
+
+def _generate_conditionals_unormalized(
+    rdm: np.ndarray, pos_array: np.ndarray, empty_orbitals: np.ndarray, marginal: float
+) -> tuple[np.ndarray, np.ndarray]:
+    """Generates the conditional and marginal probabilities for adding a particle
+    to the available empty orbitals.
```

Comment:
Docstring summary should fit on one line. Extra information can go in the main body.